OP2Helper.cpp

```diff
@@ -78,6 +78,39 @@ void CreateTubeOrWallLine(LOCATION loc1, LOCATION loc2, map_id type)
 	}
 }
 
+void RecordTubeLine(BuildingGroup& buildingGroup, LOCATION loc1, LOCATION loc2)
+{
+	// Determine edges to record along
+	const int vertEdge = loc2.x;
+	const int horizEdge = loc1.y;
+	SwapLocations(loc1, loc2);
+
+	// Record horizontal section
+	for (int x = loc1.x; x <= loc2.x; ++x) {
+		buildingGroup.RecordTube(LOCATION(x, horizEdge));
+	}
+	// Record vertical section
+	for (int y = loc1.y; y <= loc2.y; ++y) {
+		buildingGroup.RecordTube(LOCATION(vertEdge, y));
+	}
+}
+
+void RecordWallLine(BuildingGroup& buildingGroup, LOCATION loc1, LOCATION loc2, map_id wallType)

```

Comment:
Hmm, that's quite noticeable duplication here. I see there is a distinction between `RecordTube` and `RecordWall`, so that's a bit annoying to work around. Maybe with templates or lambdas, the difference can be worked around. Might even be able to share code with the `Create` methods.

On another note, I can see why you split between "Tube" and "Wall". I think I still like the idea of unpacking the wall types further though.